indextree-macros/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+use either::Either;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    punctuated::Punctuated,
+    Token,
+};
+
+#[derive(Clone, Debug)]
+struct IndexNode {
+    node: syn::Expr,
+    children: Punctuated<Self, Token![,]>,
+}
+
+impl Parse for IndexNode {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let node = input.parse::<syn::Expr>()?;
+
+        if input.parse::<Token![=>]>().is_err() {
+            return Ok(IndexNode {
+                node,
+                children: Punctuated::new(),
+            });
+        }
+
+        let children_stream;
+        syn::braced!(children_stream in input);
+        let children = children_stream.parse_terminated(Self::parse, Token![,])?;
```

Comment:
There is an implementation inefficiency in a form of recursion here. This is not much to worry about, since it depends on the structure's depth, and I doubt that depth will go over 50 in a normal scenario (we can increase recursion limit). Even so, it would be good to get rid of recursion in favour of a stack-based solution.

Reply:
This is a non-issue. Tested with nesting depth of 100 and it all works fine, I doubt anyone will every go beyond 100. (unless they're using a macro, but at this point they shouldn't be using `tree!`)